### PR TITLE
NETA DAO custom proposal module review.

### PR DESCRIPTION
**Pull request by ezekiiel**
NOT FOR MERGING.

NETA DAO is doing some custom work having a "council" that will be able to veto proposals. Opening this so we can review their changes and make sure that the first DAO to use a custom governance setup doesn't have bugs. :)

### CODE REVIEW

File: contracts/cw-proposal-single/src/contract.rs [hunk 1]
```diff
@@ -53,6 +53,11 @@ pub fn instantiate(
     let (min_voting_period, max_voting_period) =
         validate_voting_period(msg.min_voting_period, msg.max_voting_period)?;
 
+    let executor_addr: Option<Addr> = match msg.executor_addr {
+        None => None,
+        Some(addr) => Some(deps.api.addr_validate(&addr)?),
+    };
```
Comment (ezekiiel) [inline contracts/cw-proposal-single/src/contract.rs:59]:
```suggestion
    let executor_addr = msg.executor_addr.map(|human| deps.api.addr_validate(&human)).transpose()?;
```

File: contracts/cw-proposal-single/src/contract.rs [hunk 2]
```diff
@@ -383,6 +392,54 @@ pub fn execute_vote(
...
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    if config.executor_addr.is_none() {
+        return Err(ContractError::NoExecutorAssigned {});
+    }
+
+    if config.executor_addr.unwrap() != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
```
Comment (ezekiiel) [inline contracts/cw-proposal-single/src/contract.rs:410]:
```suggestion
    if config.executor_addr != Some(info.sender) {
        return Err(ContractError::Unauthorized {});
    }
```

File: contracts/cw-proposal-single/src/contract.rs [hunk 3]
```diff
@@ -587,6 +645,36 @@ pub fn execute_remove_vote_hook(
...
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    address: Option<String>,
+) -> Result<Response, ContractError> {
+    let mut config = CONFIG.load(deps.storage)?;
+
+    // Only the DAO via prop OR the executor directly may call this method.
+    if info.sender == config.dao
+        || (config.executor_addr.is_some() && config.executor_addr.unwrap() == info.sender)
```
Comment (ezekiiel) [inline contracts/cw-proposal-single/src/contract.rs:658]:
```suggestion
        || (config.executor_addr == Some(info.sender))
```

File: contracts/cw-proposal-single/src/contract.rs [hunk 4]
```diff
@@ -587,6 +645,36 @@ pub fn execute_remove_vote_hook(
...
+    {
+        // executor can be removed or reassigned to another address
+        let new_executor_addr: Option<Addr> = match address {
+            None => None,
+            Some(addr_str) => Some(deps.api.addr_validate(&addr_str).unwrap()),
+        };
```
Comment (ezekiiel) [inline contracts/cw-proposal-single/src/contract.rs:664]: Suggest doing same `.map().transpose()` maneuver as suggested above.

Reason being: generally speaking the people writing the standard library know a lot more about how to write good, high performance Rust than we do. Where we can use their stuff, we should. :)

File: contracts/cw-proposal-single/src/contract.rs [hunk 5]
```diff
@@ -596,7 +684,7 @@ pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
             query_list_proposals(deps, env, start_after, limit)
         }
         QueryMsg::ProposalCount {} => query_proposal_count(deps),
-        QueryMsg::GetVote { proposal_id, voter } => query_vote(deps, proposal_id, voter),
+        QueryMsg::Vote { proposal_id, voter } => query_vote(deps, proposal_id, voter),
```
Comment (ezekiiel) [inline contracts/cw-proposal-single/src/contract.rs:687]: @NoahSaso should they be supporting the v1 or v2 interface to be supported by the frontend? This module has a custom name, but is compatible with `cw-proposal-single`. Perhaps they should keep the name the same?

File: contracts/cw-proposal-single/src/contract.rs (hunk 4 above)
Comment (onewhiskeypls) [reply contracts/cw-proposal-single/src/contract.rs:664]: > 

good to know, ill keep that in mind next time!

File: contracts/cw-proposal-single/src/contract.rs (hunk 5 above)
Comment (onewhiskeypls) [reply contracts/cw-proposal-single/src/contract.rs:687]: Note: i branched off main before the autoclose failed execution was merged in. it included the failed proposal refund work

**Comment by JakeHartnell** [comment]
Looking good. I was writing a custom proposal module on the plane and it made me wonder if we should be making these a bit more extensible, so there is less copy pasta and code that may potentially become outdated.

We could do something similar to [cw-nfts](https://github.com/CosmWasm/cw-nfts) and use trait based extension for the Instantiate, Execute, and Query messages as well as the config. Maybe overkill?

**Comment by ezekiiel** [comment]
> @JakeHartnell: We could do something similar to [cw-nfts](https://github.com/CosmWasm/cw-nfts) and use trait based extension for the Instantiate, Execute, and Query messages as well as the config.

I actually think this is a great idea! The design there has really grown on me and I think supporting stuff like this more would be great.

**Comment by ezekiiel** [comment]
Closing for now as I think we've gotten good feedback / learning.

**Pull request closed without merging**